consensus/istanbul/backend/engine.go

```diff
@@ -505,6 +505,21 @@ func (sb *backend) Finalize(chain consensus.ChainReader, header *types.Header, s
 
 	reward.DistributeBlockReward(state, rewardSpec.Rewards)
 
+	// Only on the KIP-103 hardfork block, the following logic should be executed
+	if chain.Config().IsKIP103ForkBlock(header.Number) {
+		// rebalanceTreasury can modify the global state (state),
+		// so the existing state db should be used to apply the rebalancing result.
+		receipt, err := rebalanceTreasury(state, chain, header)
+		if err != nil {
+			logger.Error("failed to execute treasury rebalancing (KIP-103). State not changed", "err", err)
+		} else {
+			memo, err := json.Marshal(receipt)
```

Comment:
```suggestion
		}
		if receipt != nil {
			memo, err := json.Marshal(receipt)
```

Reply:
It is a possible way, but there are cons and pros to your suggestion. let me think more

Reply:
I'm sorry, but i will skip this comment. Maybe we can refine this later